BioUIMA
=======

This project uses the UIMA framework to analyze DNA sequences.  It works in two phases.  The [ProteinSequenceAnnotator](#proteinsequenceannotator) generates possible protein sequences and their possible open reading frames.  The [SequenceAlignmentAnnotator](#sequencealignmentannotator) aligns every pairwise combination of sequences from different DNA sequence inputs.  The [uima Puppet module](https://github.com/oeudtinsanmie/oeudtinsanmie-uima) may be useful to quickly set up a development environment.  However, it has currently only been sparsely tested on Ubuntu 14.04.

ProteinSequenceAnnotator
------------------------
The ProteinSequenceAnnotator generates all possible protein sequences for a given DNA sequence, using the Six-frame translation for that sequence.  Each three-RNA codon is translated into the single-letter symbol for the corresponding protein, assuming that the DNA sequences in the input are arranged in the 5'->3' direction from left to right.  Proteins that are within an [open reading frame](http://en.wikipedia.org/wiki/Open_reading_frame) are marked using capitalization of the protein symbol.  Open reading frames are also copied into a list of possible protein sequences that could be generated by this strand of DNA.  The six-frame translation and each possibly generated protein sequence for each DNA sequence are stored in a view of the CAS as a JSON hashmap. 
 
Inputs are read from the dna.txt file in the data folder of the project.  Names for sequences, denoted with #, may be placed in the same line or in a line above a given dna sequence.  If no name is provided, a numerical label is generated.

SequenceAlignmentAnnotator
--------------------------
The SequenceAlignmentAnnotator aligns every pairwise combination of sequences from different DNA sequence inputs.  Pairwise alignments of each of the six frames of each pair of DNA sequences and pairwise alignments of each possibly generated protein sequence for each pair of DNA sequences are generated and stored in a view within the CAS as a JSON hashmap.  

The pairwise alignment uses the [Wagnerâ€“Fischer algorithm](http://en.wikipedia.org/wiki/Wagner%E2%80%93Fischer_algorithm) to compute a generalization of the [Levenshtein distance](http://en.wikipedia.org/wiki/Levenshtein_distance) between the two sequences, where the costs of inserting or deleting characters is determined using the selected [Gap Penalty](http://en.wikipedia.org/wiki/Gap_penalty) method, and the cost of substituting a character may be expressed as a constant or using a [Substitution Matrix](http://compbio.pbworks.com/w/page/16252911/Pairwise%20Sequence%20Alignment%20and%20Substitution%20Matrices#ScoringMatriceswithSubstitutionPenalties).  Profile-based gap penalties require an implementation of IGapProfile to be provided to the scorer.
 
The MatchScoring configuration is read from the conf/scoring.json file. If a substitution matrix is to be used, it should be encoded as a json hashmap.  For example, the BLOSUM62 matrix would be

```
	{
		"A": {
			"A": 4,
			"R": -1,
			"N": -2,
			"D": -2,
			"C": 0,
			"Q": -1,
			"E": -1,
			"G": 0,
			"H": -2,
			"I": -1,
			"L": -1,
			"K": -1,
			"M": -1,
			"F": -2,
			"P": -1,
			"S": 1,
			"T": 0,
			"W": -3,
			"Y": -2,
			"V": 0
		},
		"R": {
			"A": -1,
			"R": 5,
			"N": 0,
			"D": -2,
			"C": -3,
			"Q": 1,
			"E": 0,
			"G": -2,
			"H": 0,
			"I": -3,
			"L": -2,
			"K": 2,
			"M": -1,
			"F": -3,
			"P": -2,
			"S": -1,
			"T": -1,
			"W": -3,
			"Y": -2,
			"V": -3
		},
		"N": {
			"A": -2,
			"R": 0,
			"N": 6,
			"D": 1,
			"C": -3,
			"Q": 0,
			"E": 0,
			"G": 0,
			"H": 1,
			"I": -3,
			"L": -3,
			"K": 0,
			"M": -2,
			"F": -3,
			"P": -2,
			"S": 1,
			"T": 0,
			"W": -4,
			"Y": -2,
			"V": -3
		},
			"D": {
			"A": -2,
			"R": -2,
			"N": 1,
			"D": 6,
			"C": -3,
			"Q": 0,
			"E": 2,
			"G": -1,
			"H": -1,
			"I": -3,
			"L": -4,
			"K": -1,
			"M": -3,
			"F": -3,
			"P": -1,
			"S": 0,
			"T": -1,
			"W": -4,
			"Y": -3,
			"V": -3
		},
		"C": {
			"A": 0,
			"R": -3,
			"N": -3,
			"D": -3,
			"C": 9,
			"Q": -3,
			"E": -4,
			"G": -3,
			"H": -3,
			"I": -1,
			"L": -1,
			"K": -3,
			"M": -1,
			"F": -2,
			"P": -3,
			"S": -1,
			"T": -1,
			"W": -2,
			"Y": -2,
			"V": -1
		},
		"Q": {
			"A": -1,
			"R": 1,
			"N": 0,
			"D": 0,
			"C": -3,
			"Q": 5,
			"E": 2,
			"G": -2,
			"H": 0,
			"I": -3,
			"L": -2,
			"K": 1,
			"M": 0,
			"F": -3,
			"P": -1,
			"S": 0,
			"T": -1,
			"W": -2,
			"Y": -1,
			"V": -2
		},
		"E": {
			"A": -1,
			"R": 0,
			"N": 0,
			"D": 2,
			"C": -4,
			"Q": 2,
			"E": 5,
			"G": -2,
			"H": 0,
			"I": -3,
			"L": -3,
			"K": 1,
			"M": -2,
			"F": -3,
			"P": -1,
			"S": 0,
			"T": -1,
			"W": -3,
			"Y": -2,
			"V": -2
		},
		"G": {
			"A": 0,
			"R": -2,
			"N": 0,
			"D": -1,
			"C": -3,
			"Q": -2,
			"E": -2,
			"G": 6,
			"H": -2,
			"I": -4,
			"L": -4,
			"K": -2,
			"M": -3,
			"F": -3,
			"P": -2,
			"S": 0,
			"T": -2,
			"W": -2,
			"Y": -3,
			"V": -3
		},
		"H": {
			"A": -2,
			"R": 0,
			"N": 1,
			"D": -1,
			"C": -3,
			"Q": 0,
			"E": 0,
			"G": -2,
			"H": 8,
			"I": -3,
			"L": -3,
			"K": -1,
			"M": -2,
			"F": -1,
			"P": -2,
			"S": -1,
			"T": -2,
			"W": -2,
			"Y": 2,
			"V": -3
		},
			"I": {
			"A": -1,
			"R": -3,
			"N": -3,
			"D": -3,
			"C": -1,
			"Q": -3,
			"E": -3,
			"G": -4,
			"H": -3,
			"I": 4,
			"L": 2,
			"K": -3,
			"M": 1,
			"F": 0,
			"P": -3,
			"S": -2,
			"T": -1,
			"W": -3,
			"Y": -1,
			"V": 3
		},
		"L": {
			"A": -1,
			"R": -2,
			"N": -3,
			"D": -4,
			"C": -1,
			"Q": -2,
			"E": -3,
			"G": -4,
			"H": -3,
			"I": 2,
			"L": 4,
			"K": -2,
			"M": 2,
			"F": 0,
			"P": -3,
			"S": -2,
			"T": -1,
			"W": -2,
			"Y": -1,
			"V": 1
		},
		"K": {
			"A": -1,
			"R": 2,
			"N": 0,
			"D": -1,
			"C": -3,
			"Q": 1,
			"E": 1,
			"G": -2,
			"H": -1,
			"I": -3,
			"L": -2,
			"K": 5,
			"M": -1,
			"F": -3,
			"P": -1,
			"S": 0,
			"T": -1,
			"W": -3,
			"Y": -2,
			"V": -2
		},
		"M": {
			"A": -1,
			"R": -1,
			"N": -2,
			"D": -3,
			"C": -1,
			"Q": 0,
			"E": -2,
			"G": -3,
			"H": -2,
			"I": 1,
			"L": 2,
			"K": -1,
			"M": 5,
			"F": 0,
			"P": -2,
			"S": -1,
			"T": -1,
			"W": -1,
			"Y": -1,
			"V": 1
		},
		"F": {
			"A": -2,
			"R": -3,
			"N": -3,
			"D": -3,
			"C": -2,
			"Q": -3,
			"E": -3,
			"G": -3,
			"H": -1,
			"I": 0,
			"L": 0,
			"K": -3,
			"M": 0,
			"F": 6,
			"P": -4,
			"S": -2,
			"T": -2,
			"W": 1,
			"Y": 3,
			"V": -1
		},
		"P": {
			"A": -1,
			"R": -2,
			"N": -2,
			"D": -1,
			"C": -3,
			"Q": -1,
			"E": -1,
			"G": -2,
			"H": -2,
			"I": -3,
			"L": -3,
			"K": -1,
			"M": -2,
			"F": -4,
			"P": 7,
			"S": -1,
			"T": -1,
			"W": -4,
			"Y": -3,
			"V": -2
		},
		"S": {
			"A": 1,
			"R": -1,
			"N": 1,
			"D": 0,
			"C": -1,
			"Q": 0,
			"E": 0,
			"G": 0,
			"H": -1,
			"I": -2,
			"L": -2,
			"K": 0,
			"M": -1,
			"F": -2,
			"P": -1,
			"S": 4,
			"T": 1,
			"W": -3,
			"Y": -2,
			"V": -2
		},
		"T": {
			"A": 0,
			"R": -1,
			"N": 0,
			"D": -1,
			"C": -1,
			"Q": -1,
			"E": -1,
			"G": -2,
			"H": -2,
			"I": -1,
			"L": -1,
			"K": -1,
			"M": -1,
			"F": -2,
			"P": -1,
			"S": 1,
			"T": 5,
			"W": -2,
			"Y": -2,
			"V": 0
		},
		"W": {
			"A": -3,
			"R": -3,
			"N": -4,
			"D": -4,
			"C": -2,
			"Q": -2,
			"E": -3,
			"G": -2,
			"H": -2,
			"I": -3,
			"L": -2,
			"K": -3,
			"M": -1,
			"F": 1,
			"P": -4,
			"S": -3,
			"T": -2,
			"W": 11,
			"Y": 2,
			"V": -3
		},
		"Y": {
			"A": -2,
			"R": -2,
			"N": -2,
			"D": -3,
			"C": -2,
			"Q": -1,
			"E": -2,
			"G": -3,
			"H": 2,
			"I": -1,
			"L": -1,
			"K": -2,
			"M": -1,
			"F": 3,
			"P": -3,
			"S": -2,
			"T": -2,
			"W": 2,
			"Y": 7,
			"V": -1
		},
		"V": {
			"A": 0,
			"R": -3,
			"N": -3,
			"D": -3,
			"C": -1,
			"Q": -2,
			"E": -2,
			"G": -3,
			"H": -3,
			"I": 3,
			"L": 1,
			"K": -2,
			"M": 1,
			"F": -1,
			"P": -2,
			"S": -2,
			"T": 0,
			"W": -3,
			"Y": -1,
			"V": 4
		}
	}
```

